lsp-mode.el

```diff
@@ -903,6 +946,220 @@ WORKSPACE is the workspace that contains the diagnostics."
                                                            (_ :note))
                                                          message)))))))
 
+(defun lsp--ht-get (tbl &rest keys)
+  "Get nested KEYS in TBL."
+  (let ((val tbl))
+    (while (and keys val)
+      (setq val (ht-get val (first keys)))
+      (setq keys (rest keys)))
+    val))
+
+;; lenses support
+
+(defun lsp--lenses-text-width (from to)
+  "Measure the width of the text between FROM and TO.
+Results are meaningful only if FROM and TO are on the same line."
+  ;; `current-column' takes prettification into account
+  (- (save-excursion (goto-char to) (current-column))
+     (save-excursion (goto-char from) (current-column))))
+
+(defun lsp--lenses-update (ov)
+  "Redraw quick-peek overlay OV."
+  (let ((offset (lsp--lenses-text-width (save-excursion
+                                          (beginning-of-visual-line)
+                                          (point))
+                                        (save-excursion
+                                          (beginning-of-line-text)
+                                          (point)))))
+    (save-excursion
+      (goto-char (overlay-start ov))
+      (overlay-put ov
+                   'before-string
+                   (format "%s%s\n"
+                           (make-string offset ?\s)
+                           (overlay-get ov 'lsp--lenses-contents))))))
+
+(defun lsp--lenses-overlay-ensure-at (pos)
+  "Find or create a lens for the line at POS."
+  (or (car (cl-remove-if-not (lambda (ov) (lsp--lenses-overlay-matches-pos ov pos)) lsp--lenses-overlays))
+      (let* ((ov (save-excursion
+                   (goto-char pos)
+                   (make-overlay (point-at-bol) (1+ (point-at-eol))))))
+        (overlay-put ov 'lsp-lens t)
+        ov)))
+
+(defun lsp--lenses-show (str pos)
+  "Show STR in an inline window at POS."
+  (let ((ov (lsp--lenses-overlay-ensure-at pos)))
+    (save-excursion
+      (goto-char pos)
+      (setf (overlay-get ov 'lsp--lenses-contents) str)
+      (lsp--lenses-update ov))
+    ov))
+
+(defun lsp--lenses-overlay-matches-pos (ov pos)
+  "Check if OV is a lens covering POS."
+  (and (overlay-get ov 'lsp-lens)
+       (<= (overlay-start ov) pos)
+       (< pos (overlay-end ov))))
+
+(defun lsp--lenses-idle-function (&optional buffer)
+  "Create iddle function for buffer BUFFER."
+  (when (or (not buffer) (eq (current-buffer) buffer))
+    (cond
+     ((not (equal (buffer-modified-tick) lsp--lenses-modified-tick))
```

Comment:
For numbers, which is more common, `=` or `eql`?

Reply:
will change it to `=`